File: algorithms/137. Single Number II/solution.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        s = 0
        for i in range(32):
            bitcount = 0
            for n in nums:
                bitcount += n & 1<<i
            if bitcount % 3:
                s += 1<<i
        # python negative uses twos-completment numbers
        if s & (1<<31):
            return s - (1<<32)
        else:
            return s

File: algorithms/137. Single Number II/solution.cc
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int single = 0, count;
        for ( int i = 0; i < 32; ++i ) {
            count = 0;
            for ( int n : nums ) {
                if ( n & (1 << i) ) {
                    ++count;
                }
            }
            if ( count % 3 ) {
                single |= 1 << i;
            }
        }
        return single;
    }
};
